wp-content/themes/currentorg/inc/business-directory-plugin.php

```diff
@@ -0,0 +1,204 @@
+
+		}
+
+	}
+
+}
+add_action( 'created_wpbdp_tag', 'wpbdp_tag_form_fields_save', 10, 2 );
+add_action( 'edited_wpbdp_tag', 'wpbdp_tag_form_fields_save', 10, 2 );
+
+/**
+ * Modify the wpdb_tag tag cloud to include wpbdp_category--id class
+ *
+ * @param Array $tag_data
+ * @return Array the new tag data
+ */
+function wpbdp_modify_tag_cloud( $tag_data ) {
+	return array_map(
+		function ( $tag ) {
+			$term = get_term( $tag['id'] );
+
+			if ( 'wpbdp_tag' === $term->taxonomy ) {
+
+				$wpbdp_tag_parent_category = get_term_meta( $tag['id'] );
+				$wpbdp_tag_parent_category = $wpbdp_tag_parent_category['wpbdp_tag_parent_category'][0];
+				$tag['class'] .= ' ' . $wpbdp_tag_parent_category;
+
+			}
+
+			return $tag;
+		},
+		(array) $tag_data
+	);
+}
+add_filter( 'wp_generate_tag_cloud_data', 'wpbdp_modify_tag_cloud' );
+
+/**
+ * Return all tags in a list in the wpbdp_tag tag cloud.
+ *
+ * Runs during an AJAX call to get the tags in the list.
+ *
+ * @param Array $args
+ */
+function wpbdp_tag_cloud_show_all_tags( $args ) {
+	
+	if( in_array( 'wpbdp_tag', $args['taxonomy'] ) ){
+
+		if ( defined( 'DOING_AJAX' ) && DOING_AJAX && isset( $_POST['action'] ) && $_POST['action'] === 'get-tagcloud') {
+
+			unset( $args['number'] );
+			$args['hide_empty'] = 0;
+
+		}
+
+	}
+
+	return $args;
+}
+add_filter( 'get_terms_args', 'wpbdp_tag_cloud_show_all_tags' );
+
+/**
+ * Show all existing tags inside of the wpbdp_tag tag cloud
+ */
+function wpbdp_tag_cloud_custom_css_js(){
+
+	global $current_screen;
+
+	if( 'wpbdp_listing' === $current_screen->post_type ) {
+		?>
+		<script type="text/javascript">
+			jQuery(window).load(function() {
+
+				jQuery("body.wp-admin #tagsdiv-wpbdp_tag #link-wpbdp_tag").trigger("click");
+				jQuery("body.wp-admin #tagsdiv-wpbdp_tag #link-wpbdp_tag").hide();
+
+			});
+
+			jQuery(document).ready(function(){
+
+				// if a category is selected/deselected, do things
+				jQuery("#wpbdp_categorychecklist input").on("click", function(){
+
+					// get id of clicked category
+					var wpbdp_category_id = jQuery(this).val();
+
+					// toggle all child tags of this category
+					jQuery("#tagsdiv-wpbdp_tag .wpbdp_category--"+wpbdp_category_id).toggleClass("hidden");
+
+					// if this categoy is unchecked, we need to remove all of its child tags that were selected
+					if(!jQuery(this).is(":checked")){
+
+						// grab text name of child tag
+						var wpbdp_parent_category_child_tag = jQuery("#tagsdiv-wpbdp_tag .wpbdp_category--"+wpbdp_category_id).text();
+
+						// loop through all selected tags and remove if it matches the name of one with an unchecked parent category
+						jQuery(".tagchecklist li").each(function(){
+							if(jQuery(this).find(".screen-reader-text").text().replace("Remove term: ", "") == wpbdp_parent_category_child_tag){
+								jQuery(this).find("button").click();
+							}
+						});
+					}
+
+				});
+
+			});
+
+			jQuery(document).ajaxStop(function() {
```

Comment:
Besides happening on ajaxStop, this should happen on page load.

Do we want to `toggleClass("hidden")`, or check/uncheck the tags?

Reply:
Also, at first I used `toggleClass`, but just switched it to `addClass("hidden")` because this is just the check to hide any tags without active parent categories when they are first loaded in. We don't need to toggle/remove the `hidden` class here because any ones that don't get hidden from this conditional will have an active parent category.

Reply:
Turns out the reason this wasn't working for me was because NoScript had untrusted the entire `.test` TLD and as a result JS wasn't running on the page. WordPress worked fine; the show/hide function didn't.

Do we want to add something to the post-save options here to make sure that the tags set on the post match the categories set on the post, so that we aren't relying on client-side code?

Reply:
> WordPress worked fine; the show/hide function didn't.

@benlk Are you referring to when the post editor loads and the tags show/hide, or when you actually select/deselect parent categories?

Reply:
But also, this code just plain doesn't appear to be running for me. I don't think the selector `#tagsdiv-wpbdp_tag .wpbdp_category--"+wpbdp_category_id` is finding any of the items. This is BDP version 5.5.7.

Reply:
> This is BDP version 5.5.7.

That might be why. I'm using BDP 5.1.9 from a direct clone of their prod site. I'll update and see what's changed.

Reply:
@benlk Also, as a note, their prod and staging sites are still on 5.1.9 as well.

Reply:
Yep, updating to 5.5.7 broke it for me as well.

Reply:
Part of why this is breaking in 5.5.7 is because somewhere between 5.1.9 and 5.5.7, the `id` for each term in the tag cloud object became the term index number instead of the actual term id, causing `$term = get_term( $tag['id'] );` to not return the proper information.

Reply:
@benlk This should work for you now. I swapped out this

```
$term = get_term( $tag[ 'id' ];

	if ( 'wpbdp_tag' === $term->taxonomy ) {

		$wpbdp_tag_parent_category = get_term_meta( $tag[ 'id' ] );
		$wpbdp_tag_parent_category = $wpbdp_tag_parent_category['wpbdp_tag_parent_category'][0];
		$tag['class'] .= ' ' . $wpbdp_tag_parent_category;

	}
```

for this:

```
$term = get_term_by( 'slug', $tag[ 'slug' ], 'wpbdp_tag' );

	if ( 'wpbdp_tag' === $term->taxonomy ) {

		$wpbdp_tag_parent_category = get_term_meta( $term->term_id );
		$wpbdp_tag_parent_category = $wpbdp_tag_parent_category['wpbdp_tag_parent_category'][0];
		$tag['class'] .= ' ' . $wpbdp_tag_parent_category;

	}
```

Reply:
Yes, this works.

Nitpicky things that we might not care about:
- [ ] The parent `li` elements for the checkboxes pick up a 6px margin-bottom from a style somewhere in `load-styles.php`, which is kinda unnecessary. 
    <img width="281" alt="Screen Shot 2019-06-25 at 14 11 34 " src="https://user-images.githubusercontent.com/1754187/60122538-668b5d80-9753-11e9-8622-1769d56cb6ba.png">
- [ ] Should we add labels to these text boxes like "These are the tags that have been applied" and "these are the tags that are available based on the selected categories"?

Reply:
> The parent li elements for the checkboxes pick up a 6px margin-bottom from a style somewhere in load-styles.php, which is kinda unnecessary.

I'll fix that since it should be fairly simple.

>  Should we add labels to these text boxes like "These are the tags that have been applied" and "these are the tags that are available based on the selected categories"?

I'll leave that up to you. Although, I don't think the default WP tag cloud displays labels like that, does it?

Reply:
I don't think the default WP tag cloud does, but it also doesn't change the tags available based on the categories in use.

Reply:
> Do we want to add something to the post-save options here to make sure that the tags set on the post match the categories set on the post, so that we aren't relying on client-side code?

@benlk I think this portion is ready for review in commit https://github.com/INN/umbrella-currentorg/pull/24/commits/14f8af3edab063f6357e52a0e1efd58fdf94d1cc.

> The parent li elements for the checkboxes pick up a 6px margin-bottom from a style somewhere in load-styles.php, which is kinda unnecessary.

>  Should we add labels to these text boxes like "These are the tags that have been applied" and "these are the tags that are available based on the selected categories"?

These are both in https://github.com/INN/umbrella-currentorg/pull/24/commits/ced8116687f09241f3908da63888e6e4af38ea58.

Reply:
LGTM.

<img width="311" alt="Screen Shot 2019-06-25 at 18 47 13 " src="https://user-images.githubusercontent.com/1754187/60138814-ac5b1c80-9779-11e9-9e65-bd4b2b1c2d8e.png">
<img width="306" alt="Screen Shot 2019-06-25 at 18 47 17 " src="https://user-images.githubusercontent.com/1754187/60138815-acf3b300-9779-11e9-8e1f-807b07a3c9f3.png">